### ISSUE TITLE
Slow Vmap when using PRNGkeys in Pytree OO implementation for dropout

### ISSUE THREAD

**Issue by niloch** [google/jax#2945]
I'm attempting to make a simple Object Oriented interface for MLP using Jax.  When following the documentation for the [stax dropout](https://jax.readthedocs.io/en/latest/_modules/jax/experimental/stax.html#Dropout) it says that a PRNGkey must be an argument in the apply_function.  All layers pass `**kwargs` to the `__call__` function so that a`key` is available for the dropout layer.

Training over batches means also passing batches of PRNGkeys to the `__call__` function of the Net.  The training cycle takes 4x longer using this implementation compared to a version without dropout (i.e. no **kwargs, no batched keys).  

Is this the right way to implement Dropout?  Am I missing some subtlety about vmap, jit, or grad that is causing this slowdown.

Any advice or feedback would be helpful.

```python3
from typing import Any, List, Tuple

import jax.numpy as np
from jax import jit, nn, random

from .base import PyTreeLike
from .tensor import Tensor

class Layer(PyTreeLike, is_abstract=True):
    """Abstract Class for Layers. Enforces subclasses to implement
    __call__, tree_flatten, tree_unflatten and registered as Pytree"""

    def __call__(self, inputs: Tensor, **kwargs) -> Tensor:
        raise NotImplementedError


class ActivationLayer(Layer, is_abstract=True):
    """Abstract Class for Activation Layers."""

    def tree_flatten(self) -> Tuple[List[None], None]:
        return ([None], None)

    @classmethod
    def tree_unflatten(cls, aux: Any, data: Iterable[Any]) -> "ActivationLayer":
        return cls()

class Linear(Layer):
    """Dense Linear Layer.
    Computes output = np.dot(w, inputs) + b"""

    def __init__(self, w: Tensor, b: Tensor) -> None:
        self.w = w
        self.b = b

    @jit
    def __call__(self, inputs: Tensor, **kwargs) -> Tensor:
        """outputs = np.dot(w, inputs) + b in single instance notation."""

        return np.dot(self.w, inputs) + self.b

    def tree_flatten(self) -> Tuple[Tuple[Tensor, Tensor], None]:
        return ((self.w, self.b), None)

    @classmethod
    def tree_unflatten(cls, aux: Any, params: Tuple[Tensor, Tensor]) -> "Linear":
        return cls(*params)


class Dropout(Layer):
    """Dropout Layer. If in train mode, keeps input activations at given probability rate,
    otherwise returns inputs directly"""

    def __init__(self, keep: float = 0.8, mode: str = "train"):
        self.keep = keep
        self.mode = mode

    @jit
    def __call__(self, inputs: Tensor, **kwargs) -> Tensor:
        """If in train mode, keeps input activations at rate,
        otherwise returns directly"""

        if self.mode == "eval":
            return inputs

        key = kwargs.get("key", None)
        if key is None:
            msg = (
                "Dropout layer requires __call__ to be called with a PRNG key "
                "argument. That is, instead of `dropout(inputs)`, call "
                "it like `dropout(inputs, key)` where `key` is a "
                "jax.random.PRNGKey value."
            )
            raise ValueError(msg)
        mask = random.bernoulli(key, self.keep, inputs.shape)
        return np.where(mask, inputs / self.keep, 0)

    def tree_flatten(self) -> Tuple[List[None], Tuple[float, str]]:
        return ([None], (self.keep, self.mode))

    @classmethod
    def tree_unflatten(cls, aux: Tuple[float, str], params: List[Any]) -> "Dropout":
        return cls(*aux)


class Tanh(ActivationLayer):
    def __call__(self, inputs: Tensor, **kwargs) -> Tensor:
        return np.tanh(inputs)

class FeedForwardNet(Layer):
    """Class for feed forward nets like Multi-layer Perceptrons."""

    def __init__(self, layers: List[Layer]) -> None:
        self.layers = layers

    @jit
    def predict(self, vector_inputs: Tensor, key: Tensor) -> Tensor:
        """Predict for a single instance by iterating over all the layers"""
        for layer in self.layers:
            vector_inputs = layer(vector_inputs, key=key)
        return vector_inputs

    @jit
    def __call__(self, batched_inputs: Tensor, batched_keys: Tensor) -> Tensor:
        """Batched Predictions"""

        return vmap(self.predict)(batched_inputs, batched_keys)

    def tree_flatten(self) -> Tuple[List[Layer], None]:
        return self.layers, None

    @classmethod
    def tree_unflatten(cls, aux: Any, params: List[Layer]) -> "FeedForwardNet":

        return cls(params, aux)

```

**Comment by jekbradbury** [comment]
If your network is very small, it's possible that generating random numbers for dropout is the bottleneck. Otherwise, can you try running with `JAX_LOG_COMPILES=1` to see if JAX is compiling anything at runtime? (That's often the culprit when things are running more slowly than expected.)

**Comment by sullivancolin** [comment]
Thank you for your help.  The compilation output looked right.   My net is indeed small for the purposes of learning from toy examples.  CPU training time vs GPU time time was initially fairly similar before implementing dropout.  I think the random.split in my training code must be the   bottleneck for the CPU as the GPU time did not change much when using dropout.  Closing

**Comment by niloch** [comment]
Thanks for your suggestion.  The compilation output looked correct, so I think the bottleneck is the random.split for generating new keys each batch. On CPU that caused training time to be quite a bit slower, but on GPU the additional time was negligible.  Closing.

**Comment by niloch** [comment]
Nevermind, the slowdown was my own fault: repeatedly casting list of keys to ndarray of keys.   I avoided creating the list in the first place and the slow down vanished.  Thanks again.
